### DISCUSSION TITLE
Unable to create a 'DbContext' of type ''.

### DISCUSSION THREAD

**Discussion by MelloRenato** [balta-io/forum#523] [Tira Dúvidas] [answered]
Prezado,

Curso: Fundamentos do Entity Framework

Estou usando .Net 8

Ao tentar gerar o migration (dotnet ef migrations add MigracaoInicial), está ocorrendo o seguinte erro:

Unable to create a 'DbContext' of type ''. The exception 'Object reference not set to an instance of an object.' was thrown while attempting to create an instance. For the different patterns supported at design time, see https://go.microsoft.com/fwlink/?linkid=851728

Vi em outros posts a necessidade de criar o construtor sem parâmetros, mas mesmo criando não resolveu.

Poderia me ajudar a identificar a causa?

public class FinancaDataContext : DbContext
{

    public DbSet<Cliente> Clientes { get; set; }
    public DbSet<Conta> Contas { get; set; }
    public DbSet<LancFinanca> LancFinancas { get; set; }

    public FinancaDataContext() { }

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) 
        => optionsBuilder.UseSqlServer("Server=localhost,1433,Database=RMFinanca;User ID=sa;Password=1q2w3e4r@#$; TrustServerCertificate=True");

    protected override void OnModelCreating(ModelBuilder modelBuilder)
    {
        modelBuilder.ApplyConfiguration(new ClienteMap());
        modelBuilder.ApplyConfiguration(new ContaMap());
        modelBuilder.ApplyConfiguration(new LancFinancaMap());
    }

}

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @MelloRenato, como vai?

Qual versão do .NET está usando? Não é a RC? Por que pelo que pesquisei tinha um bug assim que foi resolvido nas últimas versões.

Em adicional, como está a separação dos seus projetos? Onde fica a migration e onde fica o DbContext?

**Reply by MelloRenato** [reply]
Olá @andrebaltieri. Tudo bem e com você?

Estou usando a versão 8.0.103.

Sou novato com projetos Webs, então tentei montar a estrutura como você indica no curso. Segue imagem do projeto.

![image](https://github.com/balta-io/forum/assets/12654234/af378b4c-4ac5-4fa0-ad63-f98a9bfaaf0d)

Obs.: executei o clean e o  build sem erro.

**Reply by andrebaltieri** [reply]
Bom dia, @MelloRenato, como vai? Está seguindo as carreiras? É super importante seguir todas elas e não pular nenhum curso.
https://balta.io/carreiras

**Reply by MelloRenato** [reply]
Boa tarde, @andrebaltieri.

Sim estou. Já assisti até o curso "Crie aplicações Web SSR" e agora voltei para rever os cursos e fazer um primeiro programa com o que aprendi mas, parei nesse erro.

**Reply by andrebaltieri** [reply]
Qual sistema operacional (versão dele) está utilizando? Eu fiz um teste aqui no Mac e foi de boa.
Como está seu Program.cs?

**Reply by MelloRenato** [reply]
Segue os dados do SO:
Nome do Sistema Operacional	Microsoft Windows 11 Home Single Language
Versão	10.0.22621 Compilação 22621

O programs poderia causar problema nas migrations?

Segue fonte do program.cs:
using RMFinanca.Data;

var builder = WebApplication.CreateBuilder(args);
//var app = builder.Build();

//app.MapGet("/", () => "Hello World!");

using var context = new FinancaDataContext();
//{
   var ctx = new FinancaDataContext();
//    ctx.Contas.

//}


//app.Run();

**Reply by andrebaltieri** [reply]
Boa tarde, @MelloRenato, como vai?

O `Program.cs` poderia sim, mas só se estivesse carregando o contexto lá também, mas no seu caso, está apenas no seu DbContext:

```csharp
protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder) 
    => optionsBuilder.UseSqlServer("Server=localhost,1433,Database=RMFinanca;User ID=sa;Password=1q2w3e4r@#$; TrustServerCertificate=True");

```

Você pode tentar mais uma coisa... Até onde vi, este erro ocorre por que a migration não encontra um projeto executável, então temos duas saídas:

01. Apontar seu projeto web durante a migration:
```
dotnet ef --startup-project ../CAMINHO_DO_SEU_PROJETO_API/ migrations add NOME_DA_MIGRATION
```

02. Setar o projeto de API na configuração da Migração (Program.cs)
```csharp
services.AddDbContext<ApplicationDbContext>(
    options =>
        options.UseSqlServer(
            Configuration.GetConnectionString("DefaultConnection"),
            x => x.MigrationsAssembly("WebApplication1.Migrations")));
```

Fonte:
https://learn.microsoft.com/en-us/ef/core/managing-schemas/migrations/projects?tabs=dotnet-core-cli

Poderia testar e me avisar por favor?

**Reply by MelloRenato** [reply]
Bom dia @andrebaltieri.

Não compreender como criar essa as referências então preferi criar e compartilhar um repositório, assim você pode mostrar como criar essa alternativa.

https://github.com/MelloRenato/RMFinanca

Grato.
